server/server.go

```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
 	}
 	return listeners
 }
+
+// Returns true if in lame duck mode.
+func (s *Server) isLameDuckMode() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.ldm
+}
+
+// This function will close client and route listeners
+// then close clients at some interval to avoid a reconnecting storm.
+// Note that the routes are closed before so that when a client is
+// disconnected and reconnects to another server, that server does
+// not send to this server a route SUB.
+func (s *Server) lameDuckMode() {
```

Comment:
These 2 functions would be useful as well to people who embed it in other software. 

Or some other interface that achieves the lame duck shutdown like an optional flag to Shutdown()

Reply:
@ripienaar What functionality would you want to happen here from a client perspective? Maybe callback when state transitioning from lame duck to shutting down?

Reply:
@derekcollison from a client perspective. Maybe the jet lag talking but I am not sure I follow the question

As I understand this code it’s a shutdown like any other shutdown it just takes a while since it closes clients down over time nothing more? Client shouldn’t care and from the point of view of the code embedding it it should just be like Shutdown(). So something like gnatsd.Shutdown(server.LameDuckShutdown()) would do what I need quite nicely.